posts/2025-03-28-Building-Better-APIs:-Documenting-APIs-in-Open-Liberty-using-mpOpenAPI.adoc

```diff
@@ -0,0 +1,300 @@
+
+*1 . API Metadata Definitions*
+
+These annotations should used with a REST Application class.
+
+* *`@OpenAPIDefinition`* - Defines metadata for the entire API.
+
+* *`@Info`* - Provides metadata such as title, description, version, and contact details.  
+
+* *`@Server`* - Specifies one or more servers where the API is hosted.  
+
+* *`@Tag`* - Defines logical groupings of endpoints.  
+
+* *`@SecurityRequirement`* - Documentation for security (e.g., JWT authentication) of API endpoints.  
+
+[source,  java ]
+----
+@OpenAPIDefinition(
+    info = @Info(
+        title = "Blog API",
+        version = "1.0",
+        description = "This is an example API using MicroProfile OpenAPI",
+        contact = @Contact(name = "Sweetty", email = "sweetty@gmail.com"),
+        license = @License(name = "Apache 2.0", url = "https://apache.org/licenses/LICENSE-2.0.html")
+    ),
+    servers = @Server(url = "http://localhost:9081/openapi-demo", description = "Localhost Server"),
+    tags = {
+        @Tag(name = "Blog", description = "Operations related to blog management")
+    },
+    security = @SecurityRequirement(name = "jwtAuth")
+)
+public class Endpoints extends Application {
+----
+[.img_border_light]
+image::img/blog/OPENAPI-3.png[API Metadata Definitions ,width=70%,align="center"] 
+
+*2 . Security Definitions*  
+
+These annotations should used with a REST Application class.
+
+* *`@SecurityScheme`* - Defines authentication for the API.  
+
+* *`@SecuritySchemes`*- Specifies multiple security schemes if needed.  
+
+[source,  java]
+----
+@SecuritySchemes({ 
+    @SecurityScheme( 
+        securitySchemeName = "jwtAuth", 
+        description = "JWT Authentication", 
+        type = SecuritySchemeType.HTTP, 
+        scheme = "bearer", 
+        bearerFormat = "JWT" 
+    ) 
+}) 
+public class Endpoints extends Application {
+----
+[.img_border_light]
+image::img/blog/OPENAPI-4.png[Security Definitions ,width=70%,align="center"] 
+*3 . API Operations*
+
+These annotations should be used with REST resource methods.
+
+* *`@Operation`* - Describes an API operation.  
+
+* *`@APIResponse`* - Defines possible responses for an operation.  
+
+* *`@ApiResponses`* - Groups multiple @APIResponse annotations.  
+
+* *`@Content`* - Specifies the media type and structure of request/response bodies.  
+
+* *`@Schema`* - Defines the properties of a model used in API requests and responses.
+
+[source,  java]
+----
+public class Endpoints extends Application {
+@GET
+    @Path("/user/{id}")
+    @Produces(MediaType.APPLICATION_JSON)
+    @Tag(name = "Blog")
+    @Operation(summary = "Get User by ID", description = "Returns the user details for a given ID")
+    @APIResponses({
+    @APIResponse(responseCode = "200", description = "Successful Response", content = @Content(
+        mediaType = MediaType.APPLICATION_JSON, schema = @Schema(implementation = User.class)
+    )),
+    @APIResponse(responseCode = "404", description = "User not found")})
+----
```

Comment:
These annotations are attached to the method, not the class. We need to include the method signature (or at least the start of it)

Also correct the indentation.

```suggestion
[source,  java]
----
@GET
@Path("/user/{id}")
@Produces(MediaType.APPLICATION_JSON)
@Tag(name = "Blog")
@Operation(summary = "Get User by ID", description = "Returns the user details for a given ID")
@APIResponses({
@APIResponse(responseCode = "200", description = "Successful Response", content = @Content(
    mediaType = MediaType.APPLICATION_JSON, schema = @Schema(implementation = User.class)
)),
@APIResponse(responseCode = "404", description = "User not found")})
public User getUser(
----
```